main/src/main/scala/sbt/internal/server/BuildServerReporter.scala

```diff
@@ -102,95 +121,163 @@ final class BuildServerReporterImpl(
+    // Iterate through all source files 
     for {
-      source <- sources
-      filePath <- toSafePath(source)
+      source <- sources // scala
+      //id <- sourcePositionMapper(problem.position).sourcePath.toOption
+      //filePath <- toSafePath(VirtualFileRef.of(id))
     } {
-      val diagnostics = problemsByFile.getOrElse(filePath, Vector.empty)
 
-      val hadProblems = bspCompileState.hasAnyProblems.remove(filePath)
-      val shouldPublish = hadProblems || diagnostics.nonEmpty || shouldReportAllProblems
+      // twirl => twirl (scala)
+      // 1. filePath: twirl
+      // 2. problemByFiles: scala => twirl
 
-      // mark file as file with problems
-      if (diagnostics.nonEmpty) {
-        bspCompileState.hasAnyProblems.add(filePath)
-      }
+      // Get the problems associated with the current source file
+      val problems = problemsByFile.getOrElse(source, Vector.empty)
+
+      val oldDocuments = bspCompileState.hasAnyProblems.getAndUpdate(_ - source).getOrElse(source, Seq.empty)
+      // Determine if diagnostics should be published
+      // 1. The file had problems previously - we might want to update them with new ones
+      // 2. The file has fresh problems - we might want to update old ones
+      // 3. The build project is compiled for the first time - shouldReportAllProblems is set
+      val shouldPublish = oldDocuments.nonEmpty || problems.nonEmpty || shouldReportAllProblems
 
       if (shouldPublish) {
-        val params = PublishDiagnosticsParams(
-          textDocument = TextDocumentIdentifier(filePath.toUri),
-          buildTarget,
-          originId = None,
-          diagnostics,
-          reset = true
-        )
-        exchange.notifyEvent("build/publishDiagnostics", params)
+          // Group diagnostics by document
+        val diagsByDocuments = problems
+          .flatMap(mapProblemToDiagnostic)
+          .groupBy { case (document, _) => document }
+          .mapValues(_.map { case (_, diag) => diag })
+
+        val newDocuments = diagsByDocuments.keySet
+
+        bspCompileState.hasAnyProblems.updateAndGet(_ + (source -> newDocuments.toVector))
+        
+        // Iterate through both new and old documents, sending diagnostics for each
+        (newDocuments ++ oldDocuments).foreach { document =>
+          val diags: Vector[Diagnostic] = diagsByDocuments
+            .getOrElse(document, Vector.empty)
+          val params = PublishDiagnosticsParams(
+            document,
+            buildTarget,
+            originId = None,
+            diags,
+            reset = true
+          )
+          exchange.notifyEvent("build/publishDiagnostics", params)
+        }
       }
     }
   }
 
   protected override def publishDiagnostic(problem: Problem): Unit = {
     for {
       id <- problem.position.sourcePath.toOption
-      filePath <- toSafePath(VirtualFileRef.of(id))
+      (document, diagnostic) <- mapProblemToDiagnostic(problem)
```

Comment:
Add a comment to explain that mapProblemToDiagnostic use the sourcePositionMapper